### PR TITLE
On chip matrix gen

### PR DESCRIPTION
**Pull request by neworderofjamie**
This is basically a minimal, useful implementation of on-chip data generation. It supports:
- `AllToAllConnector` and `FixedProbabilityConnector`
- Delays and weights specified using `uniform` random distributions or constant values
- All formats of plastic and static synaptic matrix

To use this functionality the user needs to explicitly use PyNN SpiNNaker's NativeRNG (which I feel is a good thing as it signals no expectation of the same connectivity being generated using the numpy RNG). Irritatingly PyNN doesn't have sensible default behaviour for globally specifying the default RNG (https://github.com/NeuralEnsemble/PyNN/issues/404) so you have to do this throughout your PyNN script.

Issues #46, #47 and #48 have been created for particular missing features

wallclock

### CODE REVIEW

File: pynn_spinnaker/projections.py [hunk 1]
```diff
@@ -248,6 +251,32 @@ def _build_direct_connection(self):
 
         return direct_weights
 
+    def _estimate_max_weight(self):
```
Comment (neworderofjamie) [inline pynn_spinnaker/projections.py]: property - _max_weight_estimate

File: pynn_spinnaker/projections.py [hunk 2]
```diff
@@ -248,6 +251,32 @@ def _build_direct_connection(self):
...
+        if isinstance(weights.base_value, RandomDistribution):
+            # Get RNG and distribution
+            rng = weights.base_value.rng
+            distribution = weights.base_value.name
+            parameters = weights.base_value.parameters
+
+            # Assert that it uses our native RNG
+            assert isinstance(rng, NativeRNG)
+
+            # Update weight range based on the
```
Comment (neworderofjamie) [inline pynn_spinnaker/projections.py]: Update comment

File: pynn_spinnaker/projections.py [hunk 3]
```diff
@@ -248,6 +251,32 @@ def _build_direct_connection(self):
...
+            # estimated maximum value of distribution
+            return rng._estimate_dist_max_value(distribution, parameters)
+        # Otherwise if it's a scalar,
+        # use it to update weight range directly
```
Comment (neworderofjamie) [inline pynn_spinnaker/projections.py]: Update comment

File: pynn_spinnaker/projections.py (hunk 1 above)
Comment (neworderofjamie) [reply pynn_spinnaker/projections.py]: (or not - mundy says)

File: pynn_spinnaker/projections.py [hunk 4]
```diff
@@ -248,6 +251,32 @@ def _build_direct_connection(self):
...
+        elif isinstance(weights.base_value,
```
Comment (neworderofjamie) [inline pynn_spinnaker/projections.py]: Move into util._is_scalar (or similar)

File: pynn_spinnaker/projections.py [hunk 5]
```diff
@@ -303,3 +347,86 @@ def _directly_connectable(self):
         return (self.pre.celltype._directly_connectable and
                 self._connector._directly_connectable and
                 type(self.synapse_type) is self._static_synapse_class)
+
+    @property
+    def _generatable_on_chip(self):
```
Comment (neworderofjamie) [inline pynn_spinnaker/projections.py]: _can_generate_on_chip

File: pynn_spinnaker/projections.py [hunk 6]
```diff
@@ -303,3 +347,86 @@ def _directly_connectable(self):
...
+                    return False
+
+                # If the distribution isn't supported, return false
+                if not p.base_value.rng._supports_dist(p.base_value.name):
+                    return False
+            # Otherwise, if parameter isn't a scalar, return false
+            # **NOTE** Intuition is that parameters specified using arrays are
+            # a)Not well-defined by PyNN
+            # b)Probably wasteful to transfer to board
+            elif not isinstance(p.base_value, (int, long, np.integer,
```
Comment (neworderofjamie) [inline pynn_spinnaker/projections.py]: Again - move into utils

File: pynn_spinnaker/projections.py [hunk 7]
```diff
@@ -303,3 +347,86 @@ def _directly_connectable(self):
...
+        # If delay is random and its maximum value is
+        # larger than the maximum, return false
+        delay = s_params["delay"].base_value
+        if (isinstance(delay, RandomDistribution)
+            and delay.rng._estimate_dist_max_value(delay.name,
+                                                   delay.parameters) > max_delay):
+            return False
+
+        # If delay is a constant larger than the maximum, return false
+        if (isinstance(delay, (int, long, np.integer, float, bool))
```
Comment (neworderofjamie) [inline pynn_spinnaker/projections.py]: Again - util

File: pynn_spinnaker/projections.py [hunk 8]
```diff
@@ -303,3 +347,86 @@ def _directly_connectable(self):
...
+            assert isinstance(self._connector.rng, NativeRNG)
+
+            # Add RNG to list
+            rngs.append(self._connector.rng)
+
+        # Add any RNGs required to generate delay and weight parameters
+        rngs.extend(self._get_native_rngs("delay"))
+        rngs.extend(self._get_native_rngs("weight"))
+
+        # Return uniquified list of required RNGs
```
Comment (neworderofjamie) [inline pynn_spinnaker/projections.py]: 'with no duplicates' - just use a damn set

File: pynn_spinnaker/random.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+    # Functions to estimate the maximum value a distribution will result in
+    # **THINK** should this be moved out of NativeRNG
+    # for more general estimation of max delays etc
+    _dist_estimate_max_value = {
+        "uniform":      lambda parameters: parameters["high"],
+        "uniform_int":  lambda parameters: parameters["high"]
+    }
+
+    def __init__(self, seed=None, host_rng=None):
```
Comment (neworderofjamie) [inline pynn_spinnaker/random.py]: make host_rng non-optional

File: pynn_spinnaker/random.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+
+    # ------------------------------------------------------------------------
+    # AbstractRNG methods
+    # ------------------------------------------------------------------------
+    def next(self, n=None, distribution=None, parameters=None, mask_local=None):
+        # If a host RNG was specified, draw from that
+        if self._host_rng is not None:
+            return self._host_rng.next(n, distribution, parameters, mask_local)
+        else:
+            raise NotImplementedError("Parameters chosen using SpiNNaker "
```
Comment (neworderofjamie) [inline pynn_spinnaker/random.py]: NotImplemented is wrong exception class and not necessary if it is non-optional

File: pynn_spinnaker/random.py [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+
+    def _get_dist_param_map(self, distribution):
+        # Check translation and parameter map exists for this distribution
+        if not self._supports_dist(distribution):
+            raise NotImplementedError("SpiNNaker native RNG does not support"
+                                      "%s distributions" % distribution)
+        else:
+            return self._dist_param_maps[distribution]
+
+    def _get_dist_size(self, distribution):
```
Comment (neworderofjamie) [inline pynn_spinnaker/random.py:70]: Should check if self._supports_dist

File: pynn_spinnaker/simulator.py [hunk 12]
```diff
@@ -341,14 +356,14 @@ def _build(self, duration_ms):
             logger.debug("Found %u chip machine", len(self.system_info))
 
         # Place-and-route
+        # **NOTE** don't create application map at this stage
         logger.info("Placing and routing")
-        placements, allocations, application_map, routing_tables =\
-            place_and_route_wrapper(vertex_resources, vertex_applications,
+        placements, allocations, _, routing_tables =\
```
Comment (neworderofjamie) [inline pynn_spinnaker/simulator.py]: Could run app map be generated using wrapper?

File: pynn_spinnaker/spinnaker/regions/connection_builder.py [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+def _get_param_type_name(param):
+    # If parameter is a random distribution
+    if isinstance(param.base_value, RandomDistribution):
+        # Assert that it uses our native RNG
+        assert isinstance(param.base_value.rng, NativeRNG)
+
+        # Return distribution name
+        return param.base_value.name
+    # Otherwise if it's a scalar, return the magic string constant
+    elif isinstance(param.base_value,
```
Comment (neworderofjamie) [inline pynn_spinnaker/spinnaker/regions/connection_builder.py]: Guess what

File: pynn_spinnaker/spinnaker/regions/connection_builder.py [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+
+        # Assert that it uses our native RNG
+        assert isinstance(rng, NativeRNG)
+
+        # Return distribution size
+        rng._write_dist(fp, distribution, parameters, fixed_point)
+    # Otherwise if it's a scalar, apply fixed point scaling, round and write
+    elif isinstance(param.base_value,
+                    (int, long, np.integer, float, bool)):
+        scaled_value = round(param.base_value * (2.0 ** fixed_point))
```
Comment (neworderofjamie) [inline pynn_spinnaker/spinnaker/regions/connection_builder.py]: Rig function?

File: pynn_spinnaker/spinnaker/regions/synaptic_matrix.py [hunk 15]
```diff
@@ -227,6 +234,50 @@ def partition_matrices(self, post_vertex_slice, pre_pop_sub_rows,
...
+
+        for pre_pop, pre_n_verts in iteritems(incoming_connections):
+            # If connections from this populations
+            # should be generated on host, skip
+            if not pre_pop in pre_pop_on_chip_projection:
+                continue
+
+            # Get list of projections coming from
+            # pre_pop which should be expanded on chip
+            assert len(pre_pop_on_chip_projection[pre_pop]) == 1
```
Comment (neworderofjamie) [inline pynn_spinnaker/spinnaker/regions/synaptic_matrix.py:252]: Don't try and generate this case on-chip + add test

File: pynn_spinnaker/spinnaker/synapse_cluster.py [hunk 16]
```diff
@@ -186,6 +192,11 @@ def __init__(self, sim_timestep_ms, timer_period_us, sim_ticks,
             vert = Vertex(post_slice, receptor_index)
             for proj in synaptic_projections:
                 logger.debug("\t\t\t\tProjection:%s", proj.label)
+
+                # If this projection can be generated on chip, set flag
+                if proj._generatable_on_chip:
```
Comment (neworderofjamie) [inline pynn_spinnaker/spinnaker/synapse_cluster.py]: This is where we should check for multiple connections from one presynaptic vertex

File: pynn_spinnaker/spinnaker/synapse_cluster.py [hunk 17]
```diff
@@ -250,10 +263,24 @@ def __init__(self, sim_timestep_ms, timer_period_us, sim_ticks,
 
         logger.debug("\t\t\t%u synapse vertices", len(self.verts))
 
+        # If any matrices should be generated on chip, show message
+        if generate_matrix_on_chip:
+            # **YUCK** find connection builder executable
+            spinnaker_path = path.dirname(inspect.getfile(self.__class__))
```
Comment (neworderofjamie) [inline pynn_spinnaker/spinnaker/synapse_cluster.py]: See https://github.com/project-rig/nengo_spinnaker/blob/master/nengo_spinnaker/utils/application.py

File: pynn_spinnaker/spinnaker/regions/connection_builder.py [hunk 18]
```diff
@@ -0,0 +1,227 @@
...
+
+            logger.debug("\t\t\t\t\tWriting connection builder data for "
+                "projection key:%08x, num rows:%u, matrix type:%s, "
+                "connector type:%s, delay type:%s, weight type:%s, ",
+                prop.key, proj[1], synaptic_matrix.__name__,
+                connector.__class__.__name__, _get_param_type_name(delay),
+                _get_param_type_name(weight))
+
+            # Write header
+            fp.write(struct.pack("IIIIII", prop.key, proj[1],
```
Comment (neworderofjamie) [inline pynn_spinnaker/spinnaker/regions/connection_builder.py]: 6I

File: pynn_spinnaker/spinnaker/runtime/connection_builder/param_generator.cpp [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+//-----------------------------------------------------------------------------
+// ConnectionBuilder::ParamGenerators::Uniform
+//-----------------------------------------------------------------------------
+ConnectionBuilder::ParamGenerator::Uniform::Uniform(uint32_t *&region)
+{
+  m_Low = *reinterpret_cast<int32_t*>(region++);
+  m_Range = *reinterpret_cast<int32_t*>(region++);
+  LOG_PRINT(LOG_LEVEL_INFO, "\t\t\tUniform parameter: low:%d, range:%d",
```
Comment (neworderofjamie) [inline pynn_spinnaker/spinnaker/runtime/connection_builder/param_generator.cpp]: does python actually pass range?

**Review by neworderofjamie — COMMENTED** [review]
Mundy suggestions + things I spotted

**Comment by mundya** [comment]
Looks good to me, modulo the things marked above.

**Pull request merged**